### PR TITLE
Chore/new config

### PR DESCRIPTION
**Pull request by alexandrebodin**
This PR refactores the app configuration loading to support env variables and promote their use.

This PR is a MVC (minimum viable change) to enhance user usage without achieving the full refactoring we will be doing for the next major release. (e.g this does not change how the api / plugins config folders are loaded).

Main benefits:

- Environment variable native support
- Sync loading to make the app config available to other modules before startup
- Simplified configuration setup for users
- Promote best practices with env vars

Still some work to do: 
- [x] Update deployment doc
- [x] Update configuration documentation
- [x] Document strapi.config
- [x] check locales middleware
- [x] Refactor email conf to save in current env
- [x] Move upload plugin config to files (done in Meadia-lib)

Issues:
- Fix https://github.com/strapi/strapi/issues/5908

### CODE REVIEW

**Comment by christianguevara** [comment]
Great! I was going to do a PR pointing some places where env configuration could overwrite .json config files, which is not ideal.
For example in strapi-plugin-users-permissions/services/Jwt.js L47:49

```js
process.env.JWT_SECRET ||
        _.get(strapi.plugins['users-permissions'], 'config.jwtSecret') ||
        'oursecret'
```,

**Comment by alexandrebodin** [comment]
> Great! I was going to do a PR pointing some places where env configuration could overwrite .json config files, which is not ideal.
> For example in strapi-plugin-users-permissions/services/Jwt.js L47:49
> 
> ```js
> process.env.JWT_SECRET ||
>         _.get(strapi.plugins['users-permissions'], 'config.jwtSecret') ||
>         'oursecret'
> ```
> 
> Also i think some variables like JWT_SECRET for example, must be mandatory without defaults, like "oursecret", to avoid security issues.

We could remove the default for sure and through an error instead. We could consider removing the JWT_SECRET overwritte so the user can configure it as wished.

**Comment by christianguevara** [comment]
I think this will enforce the user to use string literals for config in the json files. Thank you for your job.

File: examples/getstarted/config/server.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+module.exports = ({ env }) => ({
+  host: env('HOST', 'localhost'),
```
Comment (Convly) [inline examples/getstarted/config/server.js:2]: Why not set it to `0.0.0.0` instead of `localhost` (in order to match [server.js](packages/strapi-generate-new/lib/resources/files/config/server.js))

File: packages/strapi-generate-new/lib/resources/json/database.json.js [hunk 2]
```diff
@@ -1,35 +1,28 @@
...
+      srv: settings.srv || false,
+      ssl: settings.ssl || false,
+    },
+    options: {
+      ...options,
+      ssl: settings.ssl || false,
+      authenticationDatabase: settings.authenticationDatabase || null,
     },
-  };
+  });
```
Comment (Convly) [inline packages/strapi-generate-new/lib/resources/json/database.json.js:27]: Writing 
```javascript
return compile({
    settings: {
      srv: false,
      ssl: false,
      ...settings,
    },
    options: {
      ssl: false,
      authenticationDatabase: null,
      ...options,
    },
  });
```
instead would reduce the extra boilerplate for default parameters. (or perhaps are you worried about the possibilities of some properties being set to nullish values in the settings?)

Comment (alexandrebodin) [reply packages/strapi-generate-new/lib/resources/json/database.json.js:27]: Yep this would be an issue with empty params (as it is coming from the cli it is fairly common to)

File: packages/strapi/lib/Strapi.js [hunk 3]
```diff
@@ -70,7 +55,9 @@ class Strapi {
     this.dir = opts.dir || process.cwd();
     this.admin = {};
     this.plugins = {};
-    this.config = this.initConfig(opts);
+    this.config = loadConfiguration(this.dir, opts);
+
+    // this.config._dump();
```
Comment (Convly) [inline packages/strapi/lib/Strapi.js:60]: 👋

File: packages/strapi-generate-new/lib/resources/files/config/server.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+  host: env('HOST', '0.0.0.0'),
+  port: env.int('PORT', 1337),
+  proxy: {
+    enabled: false,
+  },
+  cron: {
+    enabled: false,
+  },
+  admin: {
+    autoOpen: false,
```
Comment (Convly) [inline packages/strapi-generate-new/lib/resources/files/config/server.js:11]: Maybe consider using `env()` with a default value to keep consistency with `host` & `port`? (+ it would encourage the use of `env()` for this kind of configuration)


_example:_
```javascript
  proxy: {
    enabled: env.bool('PROXY_ENABLED', false),
  },
  cron: {
    enabled: env.bool('CRON_ENABLED', false),
  },
  admin: {
    autoOpen: env.bool('ADMIN_AUTO_OPEN', false),
  },
```

File: packages/strapi-generate-new/lib/resources/json/database.json.js (hunk 2 above)
Comment (Convly) [reply packages/strapi-generate-new/lib/resources/json/database.json.js:27]: Make sense 👍

File: packages/strapi-generate-new/lib/resources/files/config/server.js (hunk 4 above)
Comment (alexandrebodin) [reply packages/strapi-generate-new/lib/resources/files/config/server.js:11]: Using env vars for every variables isn't really a best practice as it adds complexity when it isn't required (e.g if you don't use cron in your app you will never need to enable it in production). It will make things harder for us to document.

File: examples/getstarted/config/server.js (hunk 1 above)
Comment (alexandrebodin) [reply examples/getstarted/config/server.js:2]: Why not yes

File: packages/strapi-plugin-email/controllers/Email.js [hunk 5]
```diff
@@ -45,17 +40,7 @@ module.exports = {
...
-      environment => {
-        return {
-          name: environment,
-          active: strapi.config.environment === environment,
-        };
-      }
-    );
-
-    ctx.send({ environments });
+    ctx.send({ environments: [] });
```
Comment (Convly) [inline packages/strapi-plugin-email/controllers/Email.js:43]: Is this still relevant to return a static empty value? Who's using this?

Comment (alexandrebodin) [reply packages/strapi-plugin-email/controllers/Email.js:43]: Nope was a temp patch :) I'm removing this completely because we are moving the email config to files.

[the remark this answers is not in the document]
Comment (alexandrebodin) [reply packages/strapi-plugin-email/controllers/Email.js:43]: Nope was a temp patch :) I'm removing this completely because we are moving the email config to files.

File: packages/strapi-plugin-upload/controllers/Upload.js [hunk 6]
```diff
@@ -99,12 +99,7 @@ module.exports = {
   },
 
   async getEnvironments(ctx) {
-    const environments = Object.keys(strapi.config.environments).map(environment => ({
-      name: environment,
-      active: strapi.config.environment === environment,
-    }));
-
-    ctx.send({ environments });
+    ctx.send({ environments: [] });
```
[the remark this answers is not in the document]
Comment (alexandrebodin) [reply packages/strapi-plugin-upload/controllers/Upload.js:102]: Nope was a temp patch :) This is already removed in the Media lib PR :)

File: packages/strapi/lib/core/bootstrap.js [hunk 7]
```diff
@@ -19,23 +19,15 @@ const pickSchema = model => {
...
 
-  // Set current environment config.
-  strapi.config.currentEnvironment = strapi.config.environments[strapi.config.environment] || {};
-
-  const defaultConnection = strapi.config.currentEnvironment.database.defaultConnection;
+  const defaultConnection = strapi.config.database.defaultConnection;
 
   // Set current connections.
-  strapi.config.connections = _.get(strapi.config.currentEnvironment, 'database.connections', {});
+  strapi.config.connections = _.get(strapi.config, 'database.connections', {});
```
[the remark this answers is not in the document]
Comment (alexandrebodin) [reply packages/strapi/lib/core/bootstrap.js:30]: I will refactor the use `config.get` has much as possible in the codebase to be consistent 👍  (which by the way just does _.get behing the scenes)

**Comment by derrickmehaffy** [comment]
@alexandrebodin looking for my review on the docs part here?

File: docs/3.0.0-beta.x/admin-panel/customization.md [hunk 8]
```diff
@@ -10,19 +10,19 @@ To apply your changes you need to [rebuild](#build) your admin panel
...
 
-**Path —** `./config/environment/**/server.json`.
-
-```json
-{
-  "host": "localhost",
-  "port": 1337,
-  "cron": {
-    "enabled": false
+**Path —** `./config/server.js`.
```
Comment (derrickmehaffy) [inline docs/3.0.0-beta.x/admin-panel/customization.md:13]: Is the per environment server.json file being depreciated? If so, it needs to go into migration guide.

File: docs/3.0.0-beta.x/admin-panel/deploy.md [hunk 9]
```diff
@@ -34,23 +34,23 @@ You might want to change the path to access to the administration panel. Here th
...
-**Path —** `./config/environment/**/server.json`.
+**Path —** `./config/server.js`.
 
 ```js
-{
-  "host": "localhost",
-  "port": 1337,
-  "cron": {
-    "enabled": false
+module.exports = {
```
Comment (derrickmehaffy) [inline docs/3.0.0-beta.x/admin-panel/deploy.md:40]: Removal of proxy block also needs to be added to migration guide (and I will need to update one-clicks for this change)

File: docs/3.0.0-beta.x/guides/deployment.md [hunk 10]
```diff
@@ -16,7 +16,7 @@ Update the `production` settings with the IP or domain name where the project wi
 
 ```js
 {
-  "host": "domain.io", // IP or domain or 0.0.0.0
+  "host": "domain.io",// IP or domain or 0.0.0.0
```
Comment (derrickmehaffy) [inline docs/3.0.0-beta.x/guides/deployment.md:19]: domain should be removed here, that isn't something that should be directly entered as it requires the OS to do DNS lookups and if the DNS fails the server will fail to bind a NaN IP address (typically gathered from reverse lookups) it also opens a vector of attack with DNS poisoning. In theory using a domain here -should- work, but it's nothing something we should recommend.

Previously it didn't matter because Strapi wasn't actually binding to anything here, but as of beta.19.4/19.5 it does matter.

File: docs/3.0.0-beta.x/installation/digitalocean-one-click.md [hunk 11]
```diff
@@ -151,19 +151,10 @@ pm2 stop strapi-develop
...
-  "port": "5432",
-  "database": "strapi",
-  "username": "strapi",
-  "password": "your-new-password"
-},
-...
+- Update the `/srv/strapi/strapi/config/.env` file with the new password.
+
+```
+DATABASE_PASSWORD=your-new-password
```
Comment (derrickmehaffy) [inline docs/3.0.0-beta.x/installation/digitalocean-one-click.md:157]: think there is a few missing variables here lol

File: docs/3.0.0-beta.x/admin-panel/customization.md (hunk 8 above)
Comment (alexandrebodin) [reply docs/3.0.0-beta.x/admin-panel/customization.md:13]: It's not deprecated per say but change. By now there is the default server.js config so I think we should only document those and explain in the config how a user can customize per env (done in the PR). 

And Yes there will be a big migration guide / script for this ;)

File: docs/3.0.0-beta.x/admin-panel/deploy.md (hunk 9 above)
Comment (alexandrebodin) [reply docs/3.0.0-beta.x/admin-panel/deploy.md:40]: Proxy is not removed in this PR. We just don't need to set it by default.

File: docs/3.0.0-beta.x/guides/deployment.md (hunk 10 above)
Comment (alexandrebodin) [reply docs/3.0.0-beta.x/guides/deployment.md:19]: Yes I agree this should be replaced with a combination of host + proxy (soon url)

File: docs/3.0.0-beta.x/installation/digitalocean-one-click.md (hunk 11 above)
Comment (alexandrebodin) [reply docs/3.0.0-beta.x/installation/digitalocean-one-click.md:157]: The other one are supposed to be in your database.js file statically. I just udpated the current doc but it needs some changes to be more env firendly. BTW does DO support seeting env vars in there manager interface ? or is it only a VPS instance ?

Comment (derrickmehaffy) [reply docs/3.0.0-beta.x/installation/digitalocean-one-click.md:157]: This is only a VPS instance, I wish we could set stuff in the management GUI would be nice, but unlikely

File: docs/3.0.0-beta.x/admin-panel/deploy.md (hunk 9 above)
Comment (derrickmehaffy) [reply docs/3.0.0-beta.x/admin-panel/deploy.md:40]: Cool, yeah got my wires crossed between this and the other PR

File: docs/3.0.0-beta.x/admin-panel/customization.md (hunk 8 above)
Comment (derrickmehaffy) [reply docs/3.0.0-beta.x/admin-panel/customization.md:13]: Gotcha :+1:

File: docs/3.0.0-beta.x/concepts/configurations.md [hunk 12]
```diff
@@ -2,54 +2,190 @@
...
+```js
+module.exports = {
+  host: '0.0.0.0',
+};
+```
+
+You can access it as
+
+```js
+strapi.config.get('server.host', 'defaultValueIfUndefined');
```
Comment (lauriejim) [inline docs/3.0.0-beta.x/concepts/configurations.md:20]: I have a question here.
Is it case sensitive?
`stra-pi.js` -> `strapi-pi.host`
`StRaPi.js` -> `StRaPi.host`

File: docs/3.0.0-beta.x/concepts/configurations.md [hunk 13]
```diff
@@ -2,54 +2,190 @@
...
+
+When using a `.js` you can either export an object:
+
+```js
+module.exports = {
+  mySecret: 'someValue',
+};
+```
+
+or a function returning a configuration object (Recommended usage). The function will get access to the [`env` utility](#casting-envrionement-variables).
```
Comment (lauriejim) [inline docs/3.0.0-beta.x/concepts/configurations.md:41]:
```diff
- Recommended
+ recommended
```

File: docs/3.0.0-beta.x/concepts/configurations.md [hunk 14]
```diff
@@ -2,54 +2,190 @@
...
+
+Instead of writting those credentials into your configuration files, you can define those variables in a `.env` file at the root of your application.
+
+**Example**
+
+```
+DATABASE_PASSWORD=acme
+```
+
+if you want to customize the path of the `.env` file to load you can set an envrionement variable called `ENV_PATH` before starting your application:
```
Comment (lauriejim) [inline docs/3.0.0-beta.x/concepts/configurations.md:63]:
```diff
- if
+ If
```

File: docs/3.0.0-beta.x/concepts/configurations.md [hunk 15]
```diff
@@ -2,54 +2,190 @@
...
+
+```sh
+$ ENV_PATH=/absolute/path/to/.env npm run start
+```
 
-These configurations are accessible through `strapi.config.favicon` and `strapi.config.public`.
+Now you can access those variables in your configuration files and application. You can use `process.env.{varName}` to access those variables anywhere.
 
-**Path —** `./config/application.json`.
+In your configuration files you will have access to a `env` utility that allows definiing defaults and casting values.
```
Comment (lauriejim) [inline docs/3.0.0-beta.x/concepts/configurations.md:71]:
```diff
- definiing
+ defining
```

File: docs/3.0.0-beta.x/concepts/configurations.md [hunk 16]
```diff
@@ -2,54 +2,190 @@
...
-    "defaultIndex": true
-  }
-}
+});
+```
+
+### Casting envrionement variables
+
+```js
+// Returns the env if defined withot casting it
```
Comment (lauriejim) [inline docs/3.0.0-beta.x/concepts/configurations.md:90]:
```diff
- withot
+ without
```

File: docs/3.0.0-beta.x/concepts/configurations.md [hunk 17]
```diff
@@ -2,54 +2,190 @@
...
+env('NAME', 'default');
+
+// Cast int (using parseInt)
```
Comment (lauriejim) [inline docs/3.0.0-beta.x/concepts/configurations.md:93]:
```diff
- Cast int
+ Cast to int
```

File: docs/3.0.0-beta.x/concepts/configurations.md [hunk 18]
```diff
@@ -2,54 +2,190 @@
...
+| `admin`                                      | Admin panel configuration                                                                                                                                   | Object        |             |
+| `admin.autoOpen`                             | Enable or disabled administration opening on start.                                                                                                         | boolean       | `true`      |
+| `admin.path`                                 | Allow to change the URL to access the admin panel.                                                                                                          | string        | `/admin`    |
+| `admin.watchIgnoreFiles`                     | Add custom files that should not be watched during development. See more [here](https://github.com/paulmillr/chokidar#path-filtering) (property `ignored`). | Array(string) | `[]`.       |
+| `admin.build`                                | Admin panel build configuration                                                                                                                             |               |             |
```
Comment (lauriejim) [inline docs/3.0.0-beta.x/concepts/configurations.md:187]: The `type` is missing.

File: docs/3.0.0-beta.x/concepts/configurations.md [hunk 19]
```diff
@@ -2,54 +2,190 @@
...
+| `proxy.ssl`                                  | Enable proxy SSL support.                                                                                                                                   | boolean       |             |
+| `proxy.host`                                 | Host name your proxy service uses for Strapi.                                                                                                               | string        |             |
+| `proxy.port`                                 | Port that your proxy service accepts connections on.                                                                                                        | integer       |             |
+| [`cron`](https://en.wikipedia.org/wiki/Cron) | Cron configuration                                                                                                                                          | Object        |             |
+| `cron.enabled`                               | Enable or disable CRON tasks to schedule jobs at specific dates.                                                                                            | boolean       | false       |
```
Comment (lauriejim) [inline docs/3.0.0-beta.x/concepts/configurations.md:182]:
```diff
- false
+ `false`
```

File: docs/3.0.0-beta.x/concepts/configurations.md [hunk 20]
```diff
@@ -2,54 +2,190 @@
...
+| `proxy.enabled`                              | Enable proxy support such as Apache or Nginx.                                                                                                               | boolean       | `false`     |
+| `proxy.ssl`                                  | Enable proxy SSL support.                                                                                                                                   | boolean       |             |
+| `proxy.host`                                 | Host name your proxy service uses for Strapi.                                                                                                               | string        |             |
+| `proxy.port`                                 | Port that your proxy service accepts connections on.                                                                                                        | integer       |             |
+| [`cron`](https://en.wikipedia.org/wiki/Cron) | Cron configuration                                                                                                                                          | Object        |             |
```
Comment (lauriejim) [inline docs/3.0.0-beta.x/concepts/configurations.md:181]: Shouldn't we link the node module we use instead?

File: docs/3.0.0-beta.x/deployment/amazon-aws.md [hunk 21]
```diff
@@ -295,32 +295,32 @@ On your development machine, navigate to your Strapi project root directory:
...
+```js
+module.exports = ({ env }) => ({
+  defaultConnection: 'default',
+  connections: {
+    default: {
+      connector: 'bookshelf',
+      settings: {
+        client: 'postgres',
+        host: env('DATABASE_HOST', '127.0.0.1'),
+        port: env.int('DATABASE_PORT', 27017),
```
Comment (lauriejim) [inline docs/3.0.0-beta.x/deployment/amazon-aws.md:313]: It's a note for all guides AND for the product.
By default, the port we set by default is `27017`.
Even if you are using Postgres, that's not the correct one.

File: docs/3.0.0-beta.x/getting-started/deployment.md [hunk 22]
```diff
@@ -90,30 +90,29 @@ Additional guides for optional software additions that compliment or improve the
 
 ### 1. Configure
 
-Update the `production` settings with the IP or domain name where the project will be running.
+We always recommend you use envrionement variables to configure your application based on the envrionement. Here is an example:
```
Comment (lauriejim) [inline docs/3.0.0-beta.x/getting-started/deployment.md:93]:
```diff
- envrionement
+ environment
```
2 times

File: docs/3.0.0-beta.x/concepts/configurations.md (hunk 20 above)
Comment (alexandrebodin) [reply docs/3.0.0-beta.x/concepts/configurations.md:181]: I just kept the previous doc but I can make the switch if you want :)

File: docs/3.0.0-beta.x/deployment/amazon-aws.md (hunk 21 above)
Comment (alexandrebodin) [reply docs/3.0.0-beta.x/deployment/amazon-aws.md:313]: This PR add a change so that the default port depends on the database, just a bad copy/paste from me here

File: docs/3.0.0-beta.x/concepts/configurations.md (hunk 12 above)
Comment (alexandrebodin) [reply docs/3.0.0-beta.x/concepts/configurations.md:20]: it is case sensitive indeed.

File: packages/strapi/lib/core/app-configuration/__tests__/env-helper.test.js [hunk 23]
```diff
@@ -0,0 +1,49 @@
+const envHelper = require('../env-helper');
+
+describe('Env helper', () => {
+  describe('env without cast', () => {
+    test('Returns undefined if not env var defined', () => {
```
Comment (Convly) [inline packages/strapi/lib/core/app-configuration/__tests__/env-helper.test.js:5]: "if there is no env var defined"

File: packages/strapi/lib/core/app-configuration/config-loader.js [hunk 24]
```diff
@@ -0,0 +1,58 @@
+'use strict';
+
+const path = require('path');
+const fs = require('fs');
+const { templateConfiguration } = require('strapi-utils');
+
+const env = require('./env-helper');
+
+module.exports = dir => {
+  if (!fs.existsSync(dir)) return {};
```
Comment (Convly) [inline packages/strapi/lib/core/app-configuration/config-loader.js:10]: It could be clearer to return undefined in order to differentiate empty conf & file not found.
imo, it should be the caller job to handle default value (such as `{}`), not the callee.

Comment (alexandrebodin) [reply packages/strapi/lib/core/app-configuration/config-loader.js:10]: This isn't about default values here but about type consistency. 
We can define the config-loader function as (string) -> Object. 

Returning undefined would break this definition and force a user to handle different return values instead of just handling an object or an exception.

If we were do consider to handle empty conf we would handle it via exceptions and errors not via return types.

Comment (Convly) [reply packages/strapi/lib/core/app-configuration/config-loader.js:10]: Although I got your point, I'm not sure hiding errors behind a default value is a good practice. So if doing `(string)?: Object` cannot be considered, throwing an exception could be a great compromise. 🙂 

PS: I might be influenced too much by TS since, based on the type checking, it's easy to spot where you don't handle potentially undefined values, which is not necessarily true in JS.

File: packages/strapi/lib/core/app-configuration/__tests__/env-helper.test.js [hunk 25]
```diff
@@ -0,0 +1,49 @@
...
+      process.env.NOT_FLOAT_VAR = '';
+      expect(envHelper.float('NOT_FLOAT_VAR')).toEqual(Number.NaN);
+    });
+
+    test('Returns a valid float when possible', () => {
+      process.env.FLOAT_VAR = '123.45';
+      expect(envHelper.float('FLOAT_VAR')).toEqual(123.45);
+    });
+  });
+});
```
Comment (petersg83) [inline packages/strapi/lib/core/app-configuration/__tests__/env-helper.test.js:49]: I would add tests for the other functions too like `bool`, `json` and `array`

File: packages/strapi/lib/core/app-configuration/config-loader.js (hunk 24 above)
Comment (alexandrebodin) [reply packages/strapi/lib/core/app-configuration/config-loader.js:10]: As Discussed more on slack :) For those interesed, In short. 

This is not considered hiding an error but just a default behavior for convenience sake (e.g rm -rf doesn't throw if folder doesn't exist) 

It would be better to do the existence check before calling the function or use a Maybe monad instead of handling undefined. Using a zero value is pretty common to :)

File: packages/strapi/lib/core/app-configuration/env-helper.js [hunk 26]
```diff
@@ -0,0 +1,58 @@
...
+
+const utils = {
+  int(...args) {
+    const value = env(...args);
+    return typeof value === 'undefined' ? value : parseInt(value, 10);
+  },
+
+  bool(...args) {
+    const value = env(...args);
+    return value === 'true';
```
Comment (petersg83) [inline packages/strapi/lib/core/app-configuration/env-helper.js:17]: if the default value is `true` (not a string), it would return false, is it a feature or a bug?

Comment (alexandrebodin) [reply packages/strapi/lib/core/app-configuration/env-helper.js:17]: Bug :) gonna fix this

File: packages/strapi/lib/core/app-configuration/env-helper.js [hunk 27]
```diff
@@ -0,0 +1,58 @@
...
+    const value = env(...args);
+    return typeof value === 'undefined' ? value : parseFloat(value);
+  },
+
+  json(key, val) {
+    const value = env(key, val);
+    try {
+      return typeof value === 'undefined' ? value : JSON.parse(value);
+    } catch (error) {
+      throw new Error(`Imposibble to parse json environment variable ${key}: ${error.message}`);
```
Comment (Convly) [inline packages/strapi/lib/core/app-configuration/env-helper.js:30]: Typo: use "Impossible" or "Unable" instead 🙂

Comment (alexandrebodin) [reply packages/strapi/lib/core/app-configuration/env-helper.js:30]: 🤦  ty!

File: packages/strapi/lib/core/app-configuration/env-helper.js [hunk 28]
```diff
@@ -2,42 +2,58 @@
...
-  float(...args) {
-    const value = env(...args);
-    return typeof value === 'undefined' ? value : parseFloat(value);
+  bool(key, defaultValue) {
+    if (!_.has(process.env, key)) {
+      return defaultValue;
+    }
+
+    const value = process.env[key];
+    return ['true', true].includes(value);
```
Comment (petersg83) [inline packages/strapi/lib/core/app-configuration/env-helper.js:34]: `true` is not necessary now since `process.env[key]` is always a string. value === 'true' would work.

**Comment by petersg83** [comment]
LGTM :)

File: docs/3.0.0-beta.x/concepts/configurations.md [hunk 29]
```diff
@@ -2,54 +2,190 @@
...
+| `proxy.ssl`              | Enable proxy SSL support.                                                                                                                                   | boolean       |             |
+| `proxy.host`             | Host name your proxy service uses for Strapi.                                                                                                               | string        |             |
+| `proxy.port`             | Port that your proxy service accepts connections on.                                                                                                        | integer       |             |
+| `cron`                   | Cron configuration (powered by [`node-schedule`](https://github.com/node-schedule/node-schedule))                                                           | Object        |             |
+| `cron.enabled`           | Enable or disable CRON tasks to schedule jobs at specific dates.                                                                                            | boolean       | false       |
```
Comment (lauriejim) [inline docs/3.0.0-beta.x/concepts/configurations.md:182]:
```diff
- false
+`false`
```

File: docs/3.0.0-alpha.x/guides/deployment.md [hunk 30]
```diff
@@ -430,19 +430,21 @@ sudo nano ecosystem.config.js
...
-    cwd: '/home/ubuntu/path-to/your-strapi-root-folder'
-    script: 'server.js',
-    env: {
-      NODE_ENV: 'production',
-      DATABASE_HOST: 'your-unique-url.rds.amazonaws.com', // database Endpoint under 'Connectivity & Security' tab
-      DATABASE_PORT: '5432',
-      DATABASE_NAME: 'strapi',  // DB name under 'Configuration' tab
-      DATABASE_USERNAME: 'postgres', // default username
-      DATABASE_PASSWORD: 'Password',
+  apps: [
```
Comment (derrickmehaffy) [inline docs/3.0.0-alpha.x/guides/deployment.md:433]: Remove Alpha doc changes

File: docs/3.0.0-beta.x/concepts/middlewares.md [hunk 31]
```diff
@@ -67,21 +67,19 @@ Every middleware will be injected into the Koa stack. To manage the load order,
 
 ## Configuration and activation
 
-To activate and configure your hook with custom options, you need to edit your `./config/environments/**/middleware.json` file in your Strapi app.
+To configure the middlewares of your application, you need to edit the `./config/middleware.js` file in your Strapi app.
```
Comment (derrickmehaffy) [inline docs/3.0.0-beta.x/concepts/middlewares.md:70]: create and edit as the file is not created by default

File: docs/3.0.0-beta.x/concepts/middlewares.md [hunk 32]
```diff
@@ -109,26 +107,109 @@ The core of Strapi embraces a small list of middlewares for performances, securi
...
+      origin: 'http://localhost',
+    },
+  },
+};
+```
+
 ### Load order
 
-The middlewares are injected into the Koa stack asynchronously. Sometimes it happens that some of these middlewares need to be loaded in a specific order. To define a load order, we created a dedicated file located in `./config/middleware.json`.
+The middlewares are injected into the Koa stack asynchronously. Sometimes it happens that some of these middlewares need to be loaded in a specific order. To define a load order, we created a dedicated file located in `./config/middleware.js`.
```
Comment (derrickmehaffy) [inline docs/3.0.0-beta.x/concepts/middlewares.md:197]: Also note about creating this file as it's not there by default

File: docs/3.0.0-beta.x/guides/client.md [hunk 33]
```diff
@@ -44,15 +44,16 @@ module.exports = strapi => {
...
-**Path —** `./config/hook.json`
-
-```json
-{
-  ...
-  "github": {
-    "enabled": true
-  }
-}
+**Path —** `./config/hook.js`
```
Comment (derrickmehaffy) [inline docs/3.0.0-beta.x/guides/client.md:47]: again a note about creating this file (believe it or not users get really confused about this type of stuff especially with extensions)

**Comment by derrickmehaffy** [comment]
Also add `.env` to the default generated `.gitignore` because people **will** forget and leak their passwords

**Comment by derrickmehaffy** [comment]
Regression of the admin auto open not being set to false after the first run. That is going to get annoying very quickly. :wink:

**Comment by derrickmehaffy** [comment]
Do we need to remove the `NODE_ENV` stuff? maybe I missed that part of the PR but the content still exists in the docs (using the local copy from this PR): http://localhost:8080/documentation/3.0.0-beta.x/getting-started/deployment.html#_2-launch-the-server

File: docs/3.0.0-beta.x/concepts/configurations.md [hunk 34]
```diff
@@ -255,270 +381,116 @@ You can find [supported database and versions](../installation/cli.html#database
...
+        schema: 'public',
       },
-      "options": {
-        "debug": true
-      }
-    }
-  }
-}
+      options: {
+        debug: true,
```
Comment (derrickmehaffy) [inline docs/3.0.0-beta.x/concepts/configurations.md:406]: Remove debug by default

File: docs/3.0.0-beta.x/concepts/configurations.md [hunk 35]
```diff
@@ -255,270 +381,116 @@ You can find [supported database and versions](../installation/cli.html#database
...
-      "options": {
-        "authenticationDatabase": "",
-        "ssl": true
-      }
-    }
-  }
-}
+      options: {
+        authenticationDatabase: '',
+        ssl: true,
```
Comment (derrickmehaffy) [inline docs/3.0.0-beta.x/concepts/configurations.md:479]: Convert over to the new env system?

File: docs/3.0.0-beta.x/concepts/middlewares.md [hunk 36]
```diff
@@ -67,21 +67,19 @@ Every middleware will be injected into the Koa stack. To manage the load order,
...
-```javascript
-{
-  ...
-  "middleware-name": {
-    "enabled": true,
-    ...
-  }
-}
-```
+**Avaialbe options**
```
Comment (derrickmehaffy) [inline docs/3.0.0-beta.x/concepts/middlewares.md:74]: Available

File: docs/3.0.0-beta.x/concepts/middlewares.md (hunk 31 above)
Comment (alexandrebodin) [reply docs/3.0.0-beta.x/concepts/middlewares.md:70]: There is a sentence just under saying the file doesn't exist by default :) will add create or edit to be safe

**Comment by abdonrd** [comment]
@alexandrebodin how to disable the `poweredBy`? I have tried with this but doesn't work:

`config/middleware.js`

```javascript
module.exports = () => ({
  settings: {
    poweredBy: {
      enable: false,
    },
  },
});
```
